src/ttlv/types.rs

```diff
@@ -0,0 +1,240 @@
+    ///
+    /// This corresponds to [`str`].
+    TextString = 0x07,
+
+    /// A TTLV byte string.
+    ///
+    /// This corresponds to `[u8]` -- a slice of bytes.
+    ByteString = 0x08,
+
+    /// A TTLV date-time.
+    ///
+    /// This is a 64-bit signed UNIX time value, in units of seconds.
+    DateTime = 0x09,
+
+    /// A TTLV interval.
+    ///
+    /// This corresponds to [`Duration`], but has a limited range and a
+    /// granularity of seconds.
+    ///
+    /// [`Duration`]: std::time::Duration
+    Interval = 0x0A,
+}
+
+impl From<TagType> for Type {
+    fn from(value: TagType) -> Self {
+        value.r#type()
+    }
+}
+
+impl fmt::Display for Type {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Type::Structure => "Structure",
+            Type::Integer => "Integer",
+            Type::LongInteger => "LongInteger",
+            Type::BigInteger => "BigInteger",
+            Type::Enumeration => "Enumeration",
+            Type::Boolean => "Boolean",
+            Type::TextString => "TextString",
+            Type::ByteString => "ByteString",
+            Type::DateTime => "DateTime",
+            Type::Interval => "Interval",
+        })
+    }
+}
+
+impl fmt::Debug for Type {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+//----------- TagType --------------------------------------------------------
+
+/// A TTLV tag-type pair.
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+pub struct TagType(u32);
+
+impl TagType {
+    /// Construct a new [`TagType`].
+    pub const fn new(tag: Tag, r#type: Type) -> Self {
+        Self(tag.0 | r#type as u32)
+    }
+
+    /// Split this into a tag-type tuple.
+    pub const fn split(self) -> (Tag, Type) {
+        (self.tag(), self.r#type())
+    }
+
+    /// The contained tag.
+    pub const fn tag(self) -> Tag {
+        Tag(self.0 & 0xFFFFFF00)
+    }
+
+    /// The contained type.
+    pub const fn r#type(self) -> Type {
+        // SAFETY: 'self.0' always contains a valid tag and type.
+        unsafe { core::mem::transmute(self.0 as u8) }
+    }
+
+    /// The value of this pair as a native-endian integer.
+    pub const fn value(self) -> u32 {
+        self.0
+    }
+
+    /// The wire-format bytes of this tag-type pair.
+    pub const fn bytes(self) -> [u8; 4] {
+        self.0.to_be_bytes()
+    }
+
+    /// Parse a tag-type pair from bytes.
+    pub const fn parse(bytes: [u8; 4]) -> Self {
+        Self(u32::from_be_bytes(bytes))
+    }
+}
+
+impl From<(Tag, Type)> for TagType {
+    fn from((tag, r#type): (Tag, Type)) -> Self {
+        Self::new(tag, r#type)
+    }
+}
+
+impl From<TagType> for (Tag, Type) {
+    fn from(value: TagType) -> Self {
+        value.split()
+    }
+}
+
+impl fmt::Display for TagType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{} {}", self.r#type(), self.tag())
```

Comment:
Why write type then tag, when this is a TagType (tag before type) and also when in TTLV the first two T's are Tag and Type, not type and tag? And maybe this would be better written as "tag (type)" to make it clearer that the type belongs to the tag?